presto-prometheus/src/main/java/io/prestosql/plugin/prometheus/PrometheusSplitManager.java

```diff
@@ -0,0 +1,285 @@
+import io.prestosql.spi.connector.FixedSplitSource;
+import io.prestosql.spi.connector.TableNotFoundException;
+import io.prestosql.spi.predicate.Domain;
+import io.prestosql.spi.predicate.Marker;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.spi.type.TimestampType;
+import org.apache.http.NameValuePair;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.message.BasicNameValuePair;
+
+import javax.inject.Inject;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static java.util.Objects.requireNonNull;
+
+public class PrometheusSplitManager
+        implements ConnectorSplitManager
+{
+    static final long OFFSET_MILLIS = 1L;
+    private final PrometheusClient prometheusClient;
+
+    @Inject
+    public PrometheusSplitManager(PrometheusClient prometheusClient)
+    {
+        this.prometheusClient = requireNonNull(prometheusClient, "client is null");
+        requireNonNull(prometheusClient.config.getQueryChunkSizeDuration(), "query-chunk-size-duration is null");
+        requireNonNull(prometheusClient.config.getMaxQueryRangeDuration(), "max-query-range-duration is null");
+    }
+
+    @Override
+    public ConnectorSplitSource getSplits(
+            ConnectorTransactionHandle transaction,
+            ConnectorSession session,
+            ConnectorTableHandle connectorTableHandle,
+            SplitSchedulingStrategy splitSchedulingStrategy)
+    {
+        PrometheusTableHandle tableHandle = (PrometheusTableHandle) connectorTableHandle;
+        PrometheusTable table = prometheusClient.getTable(tableHandle.getSchemaName(), tableHandle.getTableName());
+
+        // this can happen if table is removed during a query
+        if (table == null) {
+            throw new TableNotFoundException(tableHandle.toSchemaTableName());
+        }
+        List<ConnectorSplit> splits = new ArrayList<>();
+        generateTimesForSplits(
+                PrometheusTimeMachine.now(),
+                prometheusClient.config.getMaxQueryRangeDuration(),
+                prometheusClient.config.getQueryChunkSizeDuration(),
+                table)
+                .forEach(time -> {
+                    try {
+                        splits.add(new PrometheusSplit(buildQuery(prometheusClient.config.getPrometheusURI(),
+                                time,
+                                table.getName(),
+                                prometheusClient.config.getQueryChunkSizeDuration())));
+                    }
+                    catch (URISyntaxException e) {
```

Comment:
URI should be validated in config class validation method.

Reply:
Well it is just by virtue of the being typed URI in the config, right? But split is adding to it and it become invalid if split adds improperly so isn't this valid? Or maybe I'm not understanding the comment.

Reply:
Also `URIBuilder` in the `buildQuery` is where this URI extension to the base is happening and it throws `URISyntaxException` because that process can go awry.